File: common/src/main/java/org/sandboxpowered/silica/state/SilicaStateBuilder.java
package org.sandboxpowered.silica.state;

import com.google.common.collect.Maps;
import org.sandboxpowered.api.state.StateFactory;
import org.sandboxpowered.api.state.property.Property;
import org.sandboxpowered.api.state.property.PropertyContainer;

import java.util.Map;

public class SilicaStateBuilder<B, S extends PropertyContainer<S>> implements StateFactory.Builder<B, S> {
    private final B base;
    private final Map<String, Property<?>> properties = Maps.newHashMap();

    public SilicaStateBuilder(B base) {
        this.base = base;
    }

    public Map<String, Property<?>> getProperties() {
        return properties;
    }

    @Override
    public StateFactory.Builder<B, S> add(Property<?>... properties) {
        for (Property<?> property : properties) {
            this.properties.put(property.getName(), property);
        }

        return this;
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/state/SilicaStateBuilder.kt
package org.sandboxpowered.silica.state

import com.google.common.collect.Maps
import org.sandboxpowered.api.registry.RegistryEntry
import org.sandboxpowered.api.world.state.property.Property
import org.sandboxpowered.api.world.state.PropertyContainer
import org.sandboxpowered.api.world.state.StateProvider

class SilicaStateBuilder<B : RegistryEntry<B>, S : PropertyContainer<S>?>(private val base: B) : StateProvider.Builder<B, S> {
    private val properties: MutableMap<String, Property<*>> = Maps.newHashMap()

    fun getProperties(): Map<String, Property<*>> {
        return properties
    }

    override fun add(vararg properties: Property<*>): StateProvider.Builder<B, S> {
        for (property in properties) {
            this.properties[property.propertyName] = property
        }
        return this
    }
}